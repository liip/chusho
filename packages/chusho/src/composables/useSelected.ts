import { computed, ComputedRef, getCurrentInstance, ref, watch } from 'vue';

export type SelectedItemId = string | number;

export type UseSelected = {
  selectedItem: ComputedRef<SelectedItemId | null>;
  items: ComputedRef<SelectedItemId[]>;
  setSelectedItem: (id: SelectedItemId) => void;
  addItem: (id: SelectedItemId) => void;
  removeItem: (id: SelectedItemId) => void;
};

export default function useSelected(
  initialValue: SelectedItemId | null = null,
  propName: string | null = null
): UseSelected {
  const vm = getCurrentInstance();
  const selectedItem = ref<SelectedItemId | null>(initialValue);
  const items = ref<SelectedItemId[]>([]);

  function setSelectedItem(id: SelectedItemId) {
    selectedItem.value = id;
    if (propName) {
      vm?.emit(`update:${propName}`, selectedItem.value);
    }
  }

  function addItem(id: SelectedItemId) {
    items.value.push(id);

    if (items.value.length === 1 && !selectedItem.value) {
      selectedItem.value = id;
    }
  }

  function removeItem(id: SelectedItemId) {
    if (selectedItem.value === id) {
      selectedItem.value = items.value[0];
    }

    items.value.splice(items.value.indexOf(id), 1);
  }

  if (propName) {
    watch(
      () => vm?.props?.[propName],
      (val, oldVal) => {
        if (val !== oldVal && typeof val === 'string') {
          selectedItem.value = val;
        }
      }
    );
  }

  return {
    selectedItem: computed(() => selectedItem.value),
    items: computed(() => items.value),
    setSelectedItem,
    addItem,
    removeItem,
  };
}
