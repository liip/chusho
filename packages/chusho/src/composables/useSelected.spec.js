import { mount } from '@vue/test-utils';

import useSelected from './useSelected';

describe('useSelected', () => {
  it('initialize with default value', () => {
    const selected = useSelected('1');

    expect(selected.selectedItemId.value).toBe('1');
  });

  it('emits when there’s a component instance, a prop name and the value changes', () => {
    const wrapper = mount({
      setup() {
        return {
          selected: useSelected('1', 'selected'),
        };
      },
      mounted() {
        this.selected.setSelectedItem('2');
      },
      template: '<div></div>',
    });

    expect(wrapper.emitted()).toEqual({ 'update:selected': [['2']] });
  });

  it('updates the selectedItem when there’s a component instance and the prop value changes', async () => {
    const wrapper = mount(
      {
        props: ['modelValue'],
        setup(props) {
          return {
            selected: useSelected(props.modelValue, 'modelValue'),
          };
        },
        template: '<div></div>',
      },
      {
        props: {
          modelValue: '1',
        },
      }
    );

    expect(wrapper.vm.selected.selectedItemId.value).toEqual('1');
    await wrapper.setProps({ modelValue: '2' });
    expect(wrapper.vm.selected.selectedItemId.value).toEqual('2');
    // It should do nothing if the value didn’t actually change
    await wrapper.setProps({ modelValue: '2' });
    expect(wrapper.vm.selected.selectedItemId.value).toEqual('2');
  });

  it('setSelectedItem changes selectedItem', () => {
    const selected = useSelected('1');

    selected.setSelectedItem('2');

    expect(selected.selectedItemId.value).toBe('2');
  });

  it('addItem adds item to the list of items', () => {
    const selected = useSelected();

    selected.addItem('1');
    selected.addItem('2');
    selected.addItem('3');

    expect(selected.items.value).toMatchSnapshot();
  });

  it('removeItem removes item from the list of items', () => {
    const selected = useSelected();

    selected.addItem('1');
    selected.addItem('2');
    selected.addItem('3');

    expect(selected.items.value).toMatchSnapshot();

    selected.removeItem('2');
    expect(selected.items.value).toMatchSnapshot();

    selected.removeItem('3');
    expect(selected.items.value).toMatchSnapshot();

    selected.removeItem('1');
    expect(selected.items.value).toMatchSnapshot();
  });

  it('provides the selected item, its ID and index', () => {
    const selected = useSelected('1');

    selected.addItem('1');

    expect(selected.selectedItem.value).toEqual({ id: '1' });
    expect(selected.selectedItemId.value).toBe('1');
    expect(selected.selectedItemIndex.value).toBe(0);
  });
});
