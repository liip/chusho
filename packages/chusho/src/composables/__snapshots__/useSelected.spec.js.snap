// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSelected addItem adds item to the list of items 1`] = `
Array [
  Object {
    "id": "1",
  },
  Object {
    "id": "2",
  },
  Object {
    "id": "3",
  },
]
`;

exports[`useSelected removeItem removes item from the list of items 1`] = `
Array [
  Object {
    "id": "1",
  },
  Object {
    "id": "2",
  },
  Object {
    "id": "3",
  },
]
`;

exports[`useSelected removeItem removes item from the list of items 2`] = `
Array [
  Object {
    "id": "1",
  },
  Object {
    "id": "3",
  },
]
`;

exports[`useSelected removeItem removes item from the list of items 3`] = `
Array [
  Object {
    "id": "1",
  },
]
`;

exports[`useSelected removeItem removes item from the list of items 4`] = `Array []`;
